server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java

```diff
@@ -695,29 +718,40 @@ public void testComputeConsideringShardSizes() {
             .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey(), "99%")
             .build();
 
+        var initial = new DesiredBalance(
+            1,
+            Map.of(
+                findShardId(clusterState, "index-0"),
+                new ShardAssignment(Set.of("node-0", "node-1"), 2, 0, 0),
+                findShardId(clusterState, "index-1"),
+                new ShardAssignment(Set.of("node-0"), 1, 0, 0)
+            )
+        );
+
         var desiredBalance = new DesiredBalanceComputer(new BalancedShardsAllocator(settings)).compute(
-            DesiredBalance.INITIAL,
+            initial,
             new DesiredBalanceInput(randomInt(), routingAllocationWithDecidersOf(clusterState, clusterInfo, settings), Set.of()),
             queue(),
             input -> true
         );
 
         var resultDiskUsage = new HashMap<String, Long>();
         for (var assignment : desiredBalance.assignments().entrySet()) {
-            var nodeId = assignment.getValue().nodeIds().iterator().next();
-            var size = clusterInfo.getShardSize(assignment.getKey(), true);
-            resultDiskUsage.compute(nodeId, (k, v) -> v == null ? size : v + size);
+            for (String nodeId : assignment.getValue().nodeIds()) {
+                var size = Objects.requireNonNull(clusterInfo.getShardSize(assignment.getKey(), true));
+                resultDiskUsage.compute(nodeId, (k, v) -> v == null ? size : v + size);
+            }
         }
 
         assertThat(resultDiskUsage.values(), not(hasItems(greaterThan(1000L))));
```

Comment:
Could we (perhaps in a follow-up, not necessarily here) verify that the resulting disk usage that the balancer computes is correct too? It should sort of be the same as here, i.e., have matching remaining disk usage?